Assets/Scripts/Towers/Tower.cs

```diff
@@ -1,152 +1,207 @@
-    public float shotCooldown;
-    public float range;
-
-    private float currentShotCooldown = 0.0f;
-
-    private HashSet<Enemy> enemiesInRange;
-    private Enemy currentTarget = null;
+public class EnemyHitEvent : UnityEvent<GameObject> {}
 
-    void Start() {
-        enemiesInRange = new HashSet<Enemy>();
-        currentShotCooldown = shotCooldown;
+public class PositionHitEvent : UnityEvent<Vector3> {}
 
-        CapsuleCollider capsuleCollider = GetComponent<CapsuleCollider>();
-
-        capsuleCollider.radius = range;
-        capsuleCollider.height = range * 4.0f;
+public class Tower : MonoBehaviour {
+  public TowerType towerType;
+  
+  private float currentShotCooldown = 0.0f;
+  
+  private List<Collider> enemiesInRange;
+  private Collider currentTarget = null;
+  
+  private Vector3 launchOrigin;
+  
+  void Start() {
+    enemiesInRange = new List<Collider>();
+    currentShotCooldown = towerType.shotCooldown;
+    
+    CapsuleCollider capsuleCollider = GetComponent<CapsuleCollider>();
+    
+    capsuleCollider.radius = towerType.range;
+    capsuleCollider.height = towerType.range * 4.0f;
+    
+    launchOrigin = transform.position;
+    
+    foreach (Transform child in transform) {
+      if (child.CompareTag(TagConstants.LAUNCHORIGIN)) {
+        launchOrigin = child.position;
+        break;
+      }
     }
-
-    void Update() {
-        if (currentTarget == null && enemiesInRange.Count > 0) {
-            SelectNewTarget();
-        }
-
-        UpdateShotCooldown();
+  }
+
+  void Update() {
+    if (currentShotCooldown > 0.0f) {
+      currentShotCooldown -= Time.deltaTime;
+    } else {
+      UpdateTarget();
+      
+      if (currentTarget != null) {
+        ShootAtEnemy();
+        currentShotCooldown = towerType.shotCooldown;
+      }
     }
-
-    private void UpdateShotCooldown() {
-        if (currentShotCooldown <= 0.0f && currentTarget != null) {
-            ShootAtEnemy();
-            currentShotCooldown = shotCooldown;
+  }
+  
+  private void ShootAtEnemy() {
+    if (towerType.projectilePrefab != null) {
+      GameObject projectileObject = Instantiate(towerType.projectilePrefab, launchOrigin, Quaternion.identity);
+      Projectile projectile = projectileObject.GetComponent<Projectile>();
+      
+      if (towerType.projectileTracksEnemy) {
+        projectile.SetTargetObject(currentTarget.gameObject);
+      } else {
+        projectile.SetTargetPosition(currentTarget.gameObject.transform.position);
+      }
+      
+      projectile.SetSpeed(towerType.projectileSpeed);
+      
+      EnemyHitEvent enemyHitEvent = new EnemyHitEvent();
+      enemyHitEvent.AddListener(OnEnemyHit);
+      
+      PositionHitEvent positionHitEvent = new PositionHitEvent();
+      positionHitEvent.AddListener(OnPositionHit);
+      
+      projectile.SetEnemyHitEvent(enemyHitEvent);
+      projectile.SetPositionHitEvent(positionHitEvent);
+    }
+  }
+  
+  private void UpdateTarget() {
```

Comment:
Not sure how different most of the methods down here are to before, but they look fine to me